_posts/2021-02-01-Dynarec.md

```diff
@@ -139,11 +139,52 @@ instead of incrementing it at the end of blocks. This general pattern of
 reducing overhead on the common path, at the expense of more logic elsewhere, is
 the essence of most JIT optimizations.
 
+The remaining methods in this section I don't have direct experience
+implementing, but are worth mentioning for those emulating systems with other
+characteristics. If your address space is very large, as might be the case for
+modern systems with GBs of RAM, it may not be possible or performant to use the
+large lookup table method. Instead we can find ways to perform the expensive
+part of hash table lookup as infrequently as possible. In the most general case,
+we can make use of the fact that in most reasonably loaded hash tables, the
+first slot the key maps to contains the value we want - that is, most accesses
+do not require probing. Therefore, we can put only the code for hashing and the
+initial lookup in assembly, and exit the JIT only if probing is required. This
+is not only easier to write in assembly, but also keeps infrequently used code
+from using up cache space. It's also possible to augment the global hash table
+with a smaller, possibly block-local, hash or address table for unpredictable
```

Comment:
> possibly block-local, hash or address table

Hmmmm.......... I hadn't actually thought of doing a per-block hash-table. 

There is logic to that, each table is likely small. You could even make a dynamically sized hash function that gets back-patched and resized on misses. 

Start off with a single-entry (because a lot of indirect branches will always hit the same address) and work your way up to larger sizes. 

One downside is that you duplicate cache entries when a lot of indirect branch locations are jumping to the same set of indirect branches (a computed goto interpreter would be the worst case here, with say 256 jump-sites all indirect-jumping to the same 256 jump-sites).

There are various workarounds to this. If a local hash-table exceeds a certain size, it could be updated to point at a global hash-table. Another option is explictly detecting jumptables.